### PR TITLE
bg_ydan_sp OK

### PR DESCRIPTION
**Pull request by mzxrules**
Warning: webs may spontaneously combust.

### CODE REVIEW

File: src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c [hunk 1]
```diff
@@ -9,13 +9,23 @@ void BgYdanSp_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
+extern Gfx D_060061B0[];
+extern Gfx D_06003850[];
+extern Gfx D_06005F40[];
+
+extern CollisionHeader D_06006050;
+extern CollisionHeader D_06006460;
+
+typedef enum {
+    WEB_FLOOR,
+    WEB_WALL,
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:26]:
```suggestion
    /* 0 */ WEB_FLOOR,
    /* 1 */ WEB_WALL
```

File: src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c [hunk 2]
```diff
@@ -9,13 +9,23 @@ void BgYdanSp_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
 
-extern UNK_TYPE D_06003850;
-extern UNK_TYPE D_06005F40;
-extern UNK_TYPE D_06006050;
-extern UNK_TYPE D_060061B0;
-extern UNK_TYPE D_06006460;
+void BgYdanSp_BurnFloorWeb(BgYdanSp* thisx, GlobalContext* globalCtx);
+void BgYdanSp_FloorWebIdle(BgYdanSp* thisx, GlobalContext* globalCtx);
+void BgYdanSp_BurnWallWeb(BgYdanSp* thisx, GlobalContext* globalCtx);
+void BgYdanSp_WallWebIdle(BgYdanSp* thisx, GlobalContext* globalCtx);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:15]:
```suggestion
void BgYdanSp_BurnFloorWeb(BgYdanSp* this, GlobalContext* globalCtx);
void BgYdanSp_FloorWebIdle(BgYdanSp* this, GlobalContext* globalCtx);
void BgYdanSp_BurnWallWeb(BgYdanSp* this, GlobalContext* globalCtx);
void BgYdanSp_WallWebIdle(BgYdanSp* this, GlobalContext* globalCtx);
```

File: src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c [hunk 3]
```diff
@@ -63,29 +73,377 @@ static ColliderTrisInit D_808C09A8 = {
...
+    f32 sinsY;
+    f32 cossX;
+    f32 nSinsX;
+
+    colHeader = NULL;
+    Actor_ProcessInitChain(thisx, sInitChain);
+    this->isDestroyedSwitchFlag = thisx->params & 0x3F;
+    this->burnSwitchFlag = (thisx->params >> 6) & 0x3F;
+    thisx->params = (thisx->params >> 0xC) & 0xF;
+    DynaPolyActor_Init((DynaPolyActor*)thisx, DPM_PLAYER);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:100]: Many instances of casting `this` or `thisx` are in this file and should be cleaned up, like this one:
```suggestion
    DynaPolyActor_Init(&this->dyna, DPM_PLAYER);
```

File: src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c [hunk 4]
```diff
@@ -63,29 +73,377 @@ static ColliderTrisInit D_808C09A8 = {
...
+    DynaPoly_DeleteBgActor(globalCtx, &globalCtx->colCtx.dyna, this->dyna.bgId);
+    Collider_DestroyTris(globalCtx, &this->trisCollider);
+}
+
+void* BgYdanSp_UpdateFloorWebCollision(BgYdanSp* this) {
+    s16 newY;
+    CollisionHeader* colHeader;
+
+    colHeader = (CollisionHeader*)SEGMENTED_TO_VIRTUAL(&D_06006460);
+    colHeader->vtxList = (Vec3s*)SEGMENTED_TO_VIRTUAL(colHeader->vtxList);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:158]: We generally don't cast the result of `VIRTUAL` macros (or anything that returns a `void*` really)
```suggestion
    colHeader = SEGMENTED_TO_VIRTUAL(&D_06006460);
    colHeader->vtxList = SEGMENTED_TO_VIRTUAL(colHeader->vtxList);
```

File: src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c [hunk 5]
```diff
@@ -63,29 +73,377 @@ static ColliderTrisInit D_808C09A8 = {
...
+    Actor_ProcessInitChain(thisx, sInitChain);
+    this->isDestroyedSwitchFlag = thisx->params & 0x3F;
+    this->burnSwitchFlag = (thisx->params >> 6) & 0x3F;
+    thisx->params = (thisx->params >> 0xC) & 0xF;
+    DynaPolyActor_Init((DynaPolyActor*)thisx, DPM_PLAYER);
+    Collider_InitTris(globalCtx, &this->trisCollider);
+    Collider_SetTris(globalCtx, &this->trisCollider, thisx, &sTrisInit, this->trisColliderItems);
+    if (thisx->params == WEB_FLOOR) {
+        CollisionHeader_GetVirtual(&D_06006460, &colHeader);
+        this->actionFunc = &BgYdanSp_FloorWebIdle;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:105]: `&` should be omitted for function pointers. Relevant in the rest of the file
```suggestion
        this->actionFunc = BgYdanSp_FloorWebIdle;
```

File: src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c [hunk 6]
```diff
@@ -63,29 +73,377 @@ static ColliderTrisInit D_808C09A8 = {
...
+void BgYdanSp_BurnWeb(BgYdanSp* this, GlobalContext* globalCtx) {
+    this->timer = 30;
+    this = this;
+    func_80078884(NA_SE_SY_CORRECT_CHIME);
+    Flags_SetSwitch(globalCtx, this->isDestroyedSwitchFlag);
+    if (this->dyna.actor.params == WEB_FLOOR) {
+        this->actionFunc = &BgYdanSp_BurnFloorWeb;
+        return;
+    }
+    this->actionFunc = &BgYdanSp_BurnWallWeb;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:179]: else would be much better here imo
```suggestion
    } else {
        this->actionFunc = &BgYdanSp_BurnWallWeb;
    }
```

File: src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c [hunk 7]
```diff
@@ -63,29 +73,377 @@ static ColliderTrisInit D_808C09A8 = {
...
+    Vec3f velocity;
+    Vec3f pos2;
+    f32 distXZ;
+    f32 sins;
+    f32 coss;
+    s16 rot;
+    s16 rot2;
+    s32 i;
+
+    DECR(this->timer);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:193]: DECR shouldn't be used on its own ideally. Relevant in other functions in this file
```suggestion
    if (this->timer != 0) {
        this->timer--;
    }
```

File: src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c [hunk 8]
```diff
@@ -63,29 +73,377 @@ static ColliderTrisInit D_808C09A8 = {
...
+        for (i = 0; i < 6; i++) {
+            rot = (s16)(Rand_CenteredFloat(0x2800) + rot2);
+            sins = Math_SinS(rot);
+            coss = Math_CosS(rot);
+            pos2.x = this->dyna.actor.world.pos.x + (120.0f * sins);
+            pos2.z = this->dyna.actor.world.pos.z + (120.0f * coss);
+            distXZ = Math_Vec3f_DistXZ(&this->dyna.actor.home.pos, &pos2) * (1.0f / 120.0f);
+            if (distXZ < 0.7f) {
+                sins = Math_SinS((s16)(rot + 0x8000));
+                coss = Math_CosS((s16)(rot + 0x8000));
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:213]: I would expect a lot of the s16 casts in this file to not be necessary, but I could be wrong

**Review by fig02 — APPROVED** [review]

**Review by Roman971 — APPROVED** [review]

**Pull request merged**